component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/duo/DuoAuthenticator.java

```diff
@@ -165,54 +129,176 @@ private String getLocalAuthenticatedUser(AuthenticationContext context) {
-        JSONObject object = userInfo.getJSONObject(0);
-        phoneArray = (JSONArray) object.get(DuoAuthenticatorConstants.DUO_PHONES);
+    private void checkPhoneNumberValidation(AuthenticationContext context, String username)
+            throws AuthenticationFailedException, JSONException {
+        String mobile = getMobileClaimValue(username);
+        if (StringUtils.isNotEmpty(mobile)) {
+            JSONArray userInfo = getUserInfo(context, username);
+            context.setProperty(DuoAuthenticatorConstants.USER_INFO, userInfo);
+            JSONObject object = userInfo.getJSONObject(0);
+            JSONArray phoneArray = (JSONArray) object.get(DuoAuthenticatorConstants.DUO_PHONES);
+            if (isValidPhoneNumber(context, phoneArray, mobile)) {
+                context.setSubject(AuthenticatedUser.createLocalAuthenticatedUserFromSubjectIdentifier(username));
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("The mobile claim value and registered DUO mobile number should be in same format");
+                }
+                context.setProperty(DuoAuthenticatorConstants.NUMBER_MISMATCH, true);
+                throw new AuthenticationFailedException("Authentication failed due to mismatch in mobile numbers");
+            }
+        } else {
+            context.setProperty(DuoAuthenticatorConstants.MOBILE_CLAIM_NOT_FOUND, true);
+            throw new InvalidCredentialsException("Error while getting the mobile number from user's profile " +
+                    "for username " + username);
+        }
+    }
+
+    /**
+     * Verify the duo phone number with user's mobile claim value
+     *
+     * @param context    the authentication context
+     * @param phoneArray array with phone numbers
+     * @param mobile     the mobile claim value
+     * @return true or false
+     * @throws AuthenticationFailedException
+     * @throws JSONException
+     */
+    private boolean isValidPhoneNumber(AuthenticationContext context, JSONArray phoneArray, String mobile)
+            throws AuthenticationFailedException, JSONException {
         if (phoneArray.length() == 0) {
-            throw new AuthenticationFailedException(
-                    DuoAuthenticatorConstants.DuoErrors.ERROR_NUMBER_INVALID);
+            if (log.isDebugEnabled()) {
+                log.debug("Couldn't get the phone number of DUO user");
+            }
+            context.setProperty(DuoAuthenticatorConstants.MOBILE_NUMBER_NOT_FOUND, true);
+            throw new InvalidCredentialsException("User doesn't have a mobile number in DUO for Authentication ");
         } else {
-            return ((JSONObject) phoneArray.get(0))
-                    .getString(DuoAuthenticatorConstants.DUO_NUMBER);
+            for (int i = 0; i < phoneArray.length(); i++) {
+                if (((JSONObject) phoneArray.get(i)).getString(DuoAuthenticatorConstants.DUO_NUMBER).equals(mobile)) {
```

Comment:
Before calling this method , we have whether the mobile number is empty or not 
https://github.com/wso2-extensions/identity-outbound-auth-duo/pull/15/files#diff-318fa0251bfd40a605a65f7b90ee41c5R176